File: 02_DevApplication_IOS_avecObjectif-C/08_LacationEtLocationManagetr/08_LacationEtLocationManagetr/ViewController.m
//
//  ViewController.m
//  08_LacationEtLocationManagetr
//
//  Created by etudiant on 20/04/2017.
//  Copyright © 2017 etudiant. All rights reserved.
//

#import "ViewController.h"
@interface ViewController ()
@property (weak, nonatomic) IBOutlet UILabel *etqLatitude;
@property (weak, nonatomic) IBOutlet UILabel *etqLongitude;
@property (weak, nonatomic) IBOutlet UILabel *etqAltitude;
@property (weak, nonatomic) IBOutlet UILabel *etqPrecisionH;
@property (weak, nonatomic) IBOutlet UILabel *etqPrecisionV;
@property (weak, nonatomic) IBOutlet UILabel *etqAdresse;
@property (weak, nonatomic) IBOutlet UILabel *etqDistance;

@property (nonatomic,strong)CLLocationManager* locationManager;
@property (nonatomic, strong) CLLocation* positionCourante;
@property (nonatomic, strong) CLLocation* positionInitiale;
@property (nonatomic) CLLocationDistance distance;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.locationManager = [CLLocationManager new];
    
    //récupérer et afficher le niveau de  précision
    
    //définir le niveau de précision souhaité
    self.locationManager.desiredAccuracy = kCLLocationAccuracyBest;
    
    //je veux overrider quelques méthodes des CLLocationManager
    //pour leur donner le code dont j'ai besoin
    //Ainsi faisant, ma classe ViewController devient un délégué(execute certaines taches par procuration)
    
    //Pour obtenir cette qualité de délégué de CLLocationManager
    //pour notre classe ViewController je dois
    //- déclarer que ma classe est confotrme au protocole qui contient
    // les méthodes en question
    //-déclarer que ma classest le délégué de l'object
    //locationManager
    
    self.locationManager.delegate = self;
    //demander à l'utilisateur la permission de connaitre sa position
    [self.locationManager requestWhenInUseAuthorization];
    
    //démarrer l'activité du gestionnauire de position
    [self.locationManager startUpdatingLocation];
}
//implémenter la méthode du protocole CCLocationManagerDelegate
//qui est appelé lorsque la protion de l'utilisateur a changé
- (void)locationManager:(CLLocationManager *)manager
    didUpdateToLocation:(CLLocation *)newLocation
           fromLocation:(CLLocation *)oldLocation
{
    //récupérer et afficher la latitude
    NSString* latitude = [[NSString alloc] initWithFormat:@"%.6f",newLocation.coordinate.latitude];
    self.etqLatitude.text = latitude;
    
    //récupérer et afficher la longitude
    NSString* longitude = [[NSString alloc] initWithFormat:@"%.6f",newLocation.coordinate.longitude];
    self.etqLongitude.text = longitude;
    
    
    
    NSString* altitude= [[NSString alloc] initWithFormat:@"%.6f",newLocation.altitude];
    self.etqAltitude.text = altitude;
    
    
    NSString* precisionH = [[NSString alloc] initWithFormat:@"%.6f",newLocation.horizontalAccuracy];
    self.etqPrecisionH.text = precisionH;
    
    
    NSString* precisionV = [[NSString alloc] initWithFormat:@"%.6f",newLocation.verticalAccuracy];
    self.etqPrecisionV.text = precisionV;
    
    //je dois m'assurer qu'une position initiale a été définie
   /* if(!self.positionInitiale) //pas donnée
    {
        self.positionInitiale = newLocation;
    }*/
    
    
    self.distance += [newLocation distanceFromLocation:oldLocation];
    if(self.distance <0)
        self.distance = 0;
        
    NSString* distanceTxt = [[NSString alloc] initWithFormat:@"%.0f",_distance];
    self.etqDistance.text = distanceTxt;
    
    //récupérer (si possible) une adresse à partir de la position
    //newLocation
    CLGeocoder* geocoder = [CLGeocoder new];
    [geocoder reverseGeocodeLocation:newLocation
                   completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error) {
                       if(error)
                       {
                           NSLog(@"Geocoder a échoué avec l'erreur %@", error);
                           return;
                       }
                       if(placemarks && placemarks.count >0)
                       {
                           CLPlacemark* placemark = placemarks[0];
                           //cet objet contient aussi bien le coordonnées
                           //geographique que l'adresse (n, voie, ville, pays)
                           //récupere ces information à l'aide d'un dictionnaire
                           
                           NSDictionary* dicoAvacLadresse = [placemark addressDictionary];
                           //afficher dictionnaire avec début description
                           //pour trouver les clés
                           //NSLog(@"%@", [dicoAvacLadresse debugDescription]);
                           //la sortie nous affiche un certain nombre de clés
                           //j'en retients les suuivants
                           //street, City, ZIP, country
                           
                           
                           //récupérer à partir de ce dictionnaire
                           //les composants de l'adresse
                           NSString* adresse = [dicoAvacLadresse objectForKey:/*CNPostalAddressStreetKey*/@"Street"];
                           
                           NSString* ville = [dicoAvacLadresse objectForKey:@"City"];
                           
                           NSString* cp = [dicoAvacLadresse objectForKey:@"ZIP"];
                           
                           
                           NSString* pays = [dicoAvacLadresse objectForKey:@"Country"];
                           
                           self.etqAdresse.text =
                           [NSString stringWithFormat:@"%@ %@ %@ %@", adresse, cp,
                            ville, pays];
                           
                       }
                
                           
                   }];
    
    
    
    
}
- (IBAction)btnRAZTouched:(id)sender {
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: 03_Swift_fondamentaux/09_LocationetLocationManager/09_LocationetLocationManager/ViewController.swift
//
//  ViewController.swift
//  09_LocationetLocationManager
//
//  Created by etudiant on 12/05/2017.
//  Copyright © 2017 etudiant. All rights reserved.
//

import UIKit
import CoreLocation
class ViewController: UIViewController , CLLocationManagerDelegate{

    @IBOutlet weak var etqLatitude: UILabel!
    @IBOutlet weak var etqLongitude: UILabel!
    @IBOutlet weak var etqAltitude: UILabel!
    @IBOutlet weak var etqPrecisionH: UILabel!
    
    @IBOutlet weak var etqPrecisionV: UILabel!
    @IBOutlet weak var etqAdresse: UILabel!
    @IBOutlet weak var etqDistance: UILabel!
    
    var locationManager: CLLocationManager?
    var positionPrecedente: CLLocation?
    var distanceTotaleParcourue: CLLocationDistance = 0
    
    override func viewDidLoad() {
        super.viewDidLoad()
        etqLatitude.text = ""
        etqLongitude.text = ""
        etqAltitude.text = ""
        etqPrecisionH.text = ""
        etqAdresse.text = ""
        etqPrecisionV.text = ""
        etqDistance.text = "0"
        //récupérer une instance du service de localisation
        //qui tourne sur le dispositif
        
        locationManager = CLLocationManager()
        locationManager!.desiredAccuracy = kCLLocationAccuracyBest
        
        //lorsque le gestionnaire de position constate que la position de l'utilisateur a subit une modification
        //il appelle certaine méthode
        //imposer par le protocole CCLocationManagerDelegate
        
        //Si je veut être informé du changement de la position
        //je dois implémnter la/les methofes
        //qui m'intéressent
        //-informer le gestionnaire où se trouvent ces méthode
        locationManager!.delegate = self
        //demander l'autorisation
        locationManager!.requestWhenInUseAuthorization()
        //
        //démarrer le service de localisation
        locationManager!.startUpdatingLocation()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    //méthode du protocole CCLocationManagerDelegate
    //qui est appellé lorsque la position courante
    //a change
    //comme la position a subi, peut-être, plusiers
    //changement cette métrhode un table de positions
    //de type CCLLocation
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation])
    {
        //récupérer la position la plus récente qui esy
        //la derniere du tableau passé en paramère
        let nouvellePosition = locations[locations.count - 1]
        
        //initialiser, si besoin, la valeur de la 
        //positionPrécedent
        if(positionPrecedente == nil)
        {
            positionPrecedente = nouvellePosition
        }
        
        //afficher les coordonnées de la position
        etqAltitude.text = 	String(format: "%.2f", nouvellePosition.altitude)
        
        etqLatitude.text = String(format: "%.6f", nouvellePosition.coordinate.latitude)
        
        etqLongitude.text = String(format: "%.6f", nouvellePosition.coordinate.longitude)
        
        etqPrecisionV.text = String(format: "%.2f", nouvellePosition.verticalAccuracy)
        
        etqPrecisionH.text = String(format: "%.2f", nouvellePosition.horizontalAccuracy)
        
        distanceTotaleParcourue += nouvellePosition.distance(from: positionPrecedente!)
        
        etqDistance.text = String(format: "%0.2f", distanceTotaleParcourue)
        
        positionPrecedente = nouvellePosition
        
        //récupérer et afficher l'adresse de chaque
        //nouvelle position
        //on utilise pour cela le service distant
        //CLGeocoder
        //créer une instance de ce service
        let geocoder = CLGeocoder()
        geocoder.reverseGeocodeLocation(nouvellePosition) { (tbPlacemarks, erreur) in
            
            if erreur != nil
                 {
                print("Erreur Geocoder: \(erreur!.localizedDescription)")
                return
            }
            
            if tbPlacemarks != nil && tbPlacemarks!.count > 0
            {
                //je me limite au premier placemark du tableau
                let placemark = tbPlacemarks![0]
                
                //récupération du dico
                if let dicoAdresses = placemark.addressDictionary
                {
                    self.etqAdresse.text = "\(dicoAdresses["Street"]!)" +
                    "\n\(dicoAdresses["ZIP"]!)" +
                    "\n\(dicoAdresses["City"]!)" +
                    "\n\(dicoAdresses["Country"]!)"
                }else
                {
                    self.etqAdresse.text = ""
                }
            }
        }
        
    }

    @IBAction func btnRAZdistance(_ sender: UIButton) {
        
        distanceTotaleParcourue = 0
        etqDistance.text = "0"
    }

}

